### PR TITLE
local estimator and lenet resnet example

### PR DESCRIPTION
**Pull request by glorysdj**
This change is add local estimator for pojo style local training.
[Reviewable](https://reviewable.io/reviews/intel-analytics/analytics-zoo/1725)

### CODE REVIEW

**Comment by glorysdj** [comment]
jenkins passed: http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/2701/

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/LenetLocalEstimator.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+      val imageDirPath = params.imageDirPath
+      val batchSize = params.batchSize
+      val epoch = params.epoch
+      val threadNum = params.threadNum
+
+      val model: Module[Float] = LeNet5(10)
+      val criterion = ClassNLLCriterion[Float]()
+      val adam = new Adam[Float]()
+      val validations = Array(new Top1Accuracy[Float], new Top5Accuracy[Float], new Loss[Float])
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/LenetLocalEstimator.scala]: Doesn't need Top5

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/MnistDataLoader.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      ByteBuffer.wrap(Files.readAllBytes(Paths.get(featureFile)))
+    }
+    val labelBuffer = if (featureFile.startsWith("hdfs:")) {
+      ByteBuffer.wrap(readHdfsByte(labelFile))
+    } else {
+      ByteBuffer.wrap(Files.readAllBytes(Paths.get(labelFile)))
+    }
+    val labelMagicNumber = labelBuffer.getInt()
+
+    require(labelMagicNumber == 2049)
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/MnistDataLoader.scala]: add a message for this require

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ */
+package com.intel.analytics.zoo.examples.localEstimator
+
+import com.intel.analytics.bigdl.nn.{Contiguous, CrossEntropyCriterion, Linear, Sequential, SpatialAveragePooling, Transpose, View}
+import com.intel.analytics.bigdl.optim.{Adam, Loss, Top1Accuracy, Top5Accuracy}
+import com.intel.analytics.zoo.pipeline.api.net.TFNet
+import com.intel.analytics.zoo.pipeline.estimator.LocalEstimator
+
+object TransferLearning extends App {
+  val modelPath = "/root/glorysdj/arda-data/tf-ov-playground/working/resnet50-imagenet-fc2"
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala]: change to  a variable?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  model.add(originalModel)
+  model.add(Transpose[Float](Array((2, 4), (3, 4))))
+  model.add(Contiguous[Float]())
+  model.add(new SpatialAveragePooling[Float](2, 2, globalPooling = true))
+  model.add(new View[Float](2048).setNumInputDims(3))
+  model.add(new Linear[Float](2048, 2))
+
+  println(model)
+
+  val imageDirPath = "hdfs://172.16.0.110:9000/cifar10"
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala]: change to a variable?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    weightParameters.zip(gradParameters).foreach { case (w, g) => g.resizeAs(w) }
+    (Module.flatten[Float](weightParameters), Module.flatten[Float](gradParameters))
+  }
+
+  private def initGradWeightBias[Float](broadcastWeightBias: Array[Tensor[Float]],
+                                        localModel: Module[Float]): Unit = {
+    val (localWeightBias, localGradWeightBias) = localModel.parameters()
+    val size = localGradWeightBias.map(_.nElement()).sum
+    implicit var ev = scala.reflect.ClassTag.Float
+    val storage = Storage(size)(ev).asInstanceOf[Storage[Float]]
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala]: Storage[Float](size)?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/LenetLocalEstimator.scala (hunk 1 above)
Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/LenetLocalEstimator.scala]: ok

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/MnistDataLoader.scala (hunk 2 above)
Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/MnistDataLoader.scala]: sure

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala (hunk 4 above)
Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala]: these two will change to args

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala (hunk 5 above)
Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala]: Error:(235, 33) No ClassTag available for Float
    val storage = Storage[Float](size)
Error:(235, 33) not enough arguments for method apply: (implicit evidence$2: scala.reflect.ClassTag[Float])com.intel.analytics.bigdl.tensor.Storage[Float] in object Storage.
Unspecified value parameter evidence$2.
    val storage = Storage[Float](size)

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    val (weightParameters, gradParameters) = model.parameters()
+    require(weightParameters != null && weightParameters.length > 0,
+      s"model ${model.getName()} doesn't have any trainable parameters.")
+    require(weightParameters.size == gradParameters.size,
+      "weights and gradient number are not match")
+    weightParameters.zip(gradParameters).foreach { case (w, g) => g.resizeAs(w) }
+    (Module.flatten[Float](weightParameters), Module.flatten[Float](gradParameters))
+  }
+
+  private def initGradWeightBias[Float](broadcastWeightBias: Array[Tensor[Float]],
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala]: `initGradWeightBias[Float]` ?

Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala]: only support Float for now

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala (hunk 5 above)
Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/estimator/LocalEstimator.scala]: change to Storage[Float](size) after initGradWeightBias[Float] changed

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      val modelPath = params.modelPath
+      val inputs = params.inputs.split(",")
+      val outputs = params.outputs.split(",")
+      val imageDirPath = params.imageDirPath
+      val batchSize = params.batchSize
+      val epoch = params.epoch
+      val threadNum = params.threadNum
+
+      val originalModel = TFNet.fromSavedModel(modelPath, inputs, outputs)
+      println(originalModel)
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala]: debug code?  you have two print model

Comment (glorysdj) [reply zoo/src/main/scala/com/intel/analytics/zoo/examples/localEstimator/TransferLearning.scala]: yes, print model before and after transfer learning

**Comment by qiuxin2012** [comment]
public method should have scala doc.

**Comment by glorysdj** [comment]
> public method should have scala doc.

added

**Comment by glorysdj** [comment]
http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/2752

**Pull request merged**